library.json

```diff
@@ -18,16 +18,9 @@
   "platforms": ["espressif8266", "espressif32"],
   "dependencies": [
     {
-      "owner": "me-no-dev",
-      "name": "ESPAsyncTCP",
-      "version": "^1.2.2",
-      "platforms": "espressif8266"
-    },
-    {
-      "owner": "me-no-dev",      
-      "name": "AsyncTCP",
-      "version": "^1.1.1",
-      "platforms": "espressif32"
+     "owner": "yubox-node-org",
+     "name": "AsyncTCPSock",
+     "version": "https://github.com/yubox-node-org/AsyncTCPSock"
```

Comment:
Hello, I'm curious to understand how this library compare to the ESPHome maintained fork (https://github.com/esphome/AsyncTCP) and what it brings compare to AsyncTCP ? 
I am using a fork of `yubox-node-org/ESPAsyncWebServer` also, but with the ESPHome maintained AsyncTCP.
Thanks!

Reply:
Who knows... Do not see much activity in this fork recently.
How is it works `yubox-node-org/ESPAsyncWebServer` with `esphome/AsyncTCP`?
Is it more stable? My impression was that `yubox-node-org/ESPAsyncWebServer`  must be used with it's own version of `AsyncTCPSock`.
Have you tried `esphome/ESPAsyncWebServer`? They have added some fixes there, but yubox fork is almost a full rework.

Reply:
FYI I am maintaining a more up to date version of the `yubox-node-org` fork, which is deployed in [PlatformIO registry](https://registry.platformio.org/libraries/mathieucarbou/ESP%20Async%20WebServer) and Arduino registry. 

- It is compatible, but cleaned up regarding CI, spdiff editor, etc. 
- It is also compatible  Arduino Json 7
- It contains the few improvements made by ESPHome

The ESPHome folks decided to fork the original repo, which is IMO wrong because the yubox-node-org introduces a LOT of fixes regarding concurrency issues when using tasks especially on dual core systems. So they miss all that, especially on the WebSocket part.

I know a lot of projects depending on the `yubox-node-org` fork instead of the ESPHome one or original one for this reason. It adds a lot of stability and prevents corrupted heap causes by concurrent access to queues in the lib.

The version I maintain is using `esphome/AsyncTCP` because the ESPHome team does a great job IMO to maintain it. They fixed a couple of issues and introduced IPv6.

Reply:
Thanks! I'll take a look into your fork also! If you like I can rebase this PR onto your fork.
I also have some improvements for async webserver that I can share.

Reply:
> Thanks! I'll take a look into your fork also! If you like I can rebase this PR onto your fork. I also have some improvements for async webserver that I can share.

Yes! An adapted version of this change would definitely be interesting because sadly the api change in the buffer is not compatible with the original API. But making is compatible is really harder I think.

For the moment I have to workaround that by using feature detection like here:

https://github.com/ayushsharma82/ESP-DASH/pull/195/files#diff-b22c24b3d761e54f8997b5313d563e3b2b58217a99f7e9e95beb68ad19ba6e13R343-R351

The problem is that the original API returns a pointer to a buffer, which requires it to be responsible of its destruction:

https://github.com/esphome/ESPAsyncWebServer/blob/master/src/AsyncWebSocket.h#L333-L334

The change using yubox fork with a shared ptr allows a buffer created by the caller to still be referenced until all the clients have used it. His change is well explained in his commit here:

https://github.com/me-no-dev/ESPAsyncWebServer/commit/4963ce9da983e433d938e8b978327f0eac2ddd8c

I don't know how to make the original API compatible with the use of shared ptr.

Reply:
I have opened a PR in my fork to add this API back.
https://github.com/mathieucarbou/ESPAsyncWebServer/pull/3

Reply:
> The AsyncTCPSock solves the deadlock problem by removing the entire event queue, and instead using the high-level socket interface to implement the same API as AsyncTCP. Full explanation [here](https://github.com/me-no-dev/ESPAsyncWebServer/issues/825#issuecomment-748538037).

Thanks for the clarification about the API compatibility! That makes sense.

> I cannot comment fully on whether the ESPHome fork of AsyncTCP has addressed the deadlock issue, but from a cursory inspection, it appears that it still uses an event queue therefore could also deadlock like the original library.

Noted! Have to take a closer look there also. Appreciate your work on this!

Reply:
On my side I've already tried to switch to AsyncTCPSock and beside a little API change regarding the connected() vs disconnected(), it was not working. I am using ESPAsyncWebSwevee but also bertmelis-espMqttClient which makes use of AsyncTCP directly. I saw no crash but tasks seems to be locked. I didn't further inspect to not loose time but this is not a drop-in replacement.

Reply:
> The ESPHome folks decided to fork the original repo, which is IMO wrong because the yubox-node-org introduces a LOT of fixes regarding concurrency issues when using tasks especially on dual core systems. So they miss all that, especially on the WebSocket part.
> 
> I know a lot of projects depending on the `yubox-node-org` fork instead of the ESPHome one or original one for this reason. It adds a lot of stability and prevents corrupted heap causes by concurrent access to queues in the lib.
> 
> The version I maintain is using `esphome/AsyncTCP` because the ESPHome team does a great job IMO to maintain it. They fixed a couple of issues and introduced IPv6.

For what it's worth, I've been using both yubox-node-org's AsyncWS and AsyncTCP together happily in my project for a number of years now, with no issues. It's rock-solid I have probably a few thousand deployments out there. I did try a port to https://github.com/hoeken/PsychicHTTP which uses IDF's espwebserver, with multi-core lwip but it didn't beat the performance and memory footprint of my old libs.

Reply:
@proddy : funny, exactly in the same situation! Using this fork also, plus AsyncTCP (but from ESPHome since more up to date) and also using the same mqtt client. Apps have a huge uptime, no memory leak, and small footprint also, as soon as we correctly configure the stack size of each task. 

That's the reasons I decided to fork and maintain it here: https://registry.platformio.org/libraries/mathieucarbou/ESP%20Async%20WebServer (also avail. for Arduino). There were a couple of things missing that I added like:
- Arduino Json 7 support
- A few improvements from ESPHome team regarding stack and queue config
- `AsyncWebSocketMessageBuffer` and `makeBuffer()` added to be API-compatible with the original fork from me-no-dev
- etc ;-)

Reply:
thanks @mathieucarbou - I saw your makeBuffer() change which is sweet. I'll try and your forks and see how they compare. nice work :)

Reply:
I did a quick test and swapping out my AsyncWS (yubox forked and also ArduinoJson7-d) with your version and esphome's AsyncTCP. Heap has increased by 9kb and max alloc too slightly so next I'm going to deep dive and find out what is causing the additional memory usage.

Reply:
I am wondering if this is caused by the new AsyncTCP from esphome and IPv6 support. I notice a slight increase too but not as much.
My fork from younodebox has only a few changes you can verify easily with a diff and none of them impact heap / stack usage (the async message buffer wrapper ust allocates and then deallocates).

Reply:
Let me know about your findings! I am interested in improving that if possible.

Reply:
@proddy : did you find out ? is it possible that this increase is caused by the difference values for the queue sizes ?

Reply:
to be honest I can't remember! I'm pretty sure it's in AsyncTCP. I'm still using the ESPHome fork with IPv6 support with CONFIG_ASYNC_TCP_STACK_SIZE set to 5120 which is enough for ESP32 (2.3k) and ESP32S3 (3.5k)

Reply:
> to be honest I can't remember! I'm pretty sure it's in AsyncTCP. I'm still using the ESPHome fork with IPv6 support with CONFIG_ASYNC_TCP_STACK_SIZE set to 5120 which is enough for ESP32 (2.3k) and ESP32S3 (3.5k)

Ok. Because I ran some tests this morning:

![Screenshot 2024-06-04 at 10 54 27](https://github.com/yubox-node-org/ESPAsyncWebServer/assets/61346/3bd4ae74-50fb-4bc2-9428-95eb02a6b24e)

* 9h57 - 10h05: AsyncTCPSock with default stack size
* 10h06 - 10h14 : AsyncTCPSock with 4k stack size
* Before and after: AsyncTCP with 4k stack size.

If you look from 10h35 and after: this is typically what I like to see. There is no big allocations / deallocation on heap, which avoids fragmentation. This is quite stable.

Spikes are app reloads (ESP-DASH PRo + WebSerial Pro). 
This app also has mqtt publications each 5 seconds, dashboard refresh each 500ms, and log streamed to console (debug mode).

Reply:
I am running with my forks:
- https://github.com/mathieucarbou/AsyncTCP
- https://github.com/mathieucarbou/ESPAsyncWebServer

My AsyncTCP fork is based on ESPHome, but supports Arduino 3, Ipv6 for Arduino 3 also, and fixes some issus in the ESPHome implementation of Ipv6 (I've PR-ed them if I remember). It also include a workaround of a bug introduced in Arduino 3 in the IpAddress implementation.

Reply:
> I am running with my forks:
> 
> * https://github.com/mathieucarbou/AsyncTCP
> * https://github.com/mathieucarbou/ESPAsyncWebServer
> 
> My AsyncTCP fork is based on ESPHome, but supports Arduino 3, Ipv6 for Arduino 3 also, and fixes some issus in the ESPHome implementation of Ipv6 (I've PR-ed them if I remember). It also include a workaround of a bug introduced in Arduino 3 in the IpAddress implementation.

nice. I'll do some benchmarking too. Having local copies of the libraries is a nightmare to maintain so it'll be good if I switch to public libraries.